### PR TITLE
Tor, Angelica, Hannah - SuperMarket Betsy - Octos

### PR DESCRIPTION
**Pull request by hannahlcameron**
# bEtsy
Congratulations! You're submitting your assignment! These comprehension questions should be answered by all members of your team, not by a single teammate.

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did your team break up the work to be done? | We broke up work assignments evenly by models (2 each) and then worked our way through the controllers as we needed them. We continued to assign work based off of priority and difficulty of task. We worked mostly independently, but closely coordinated so that we were in sync with each other.
How did your team utilize git to collaborate? | We used git to collaborate by leveraging the ability to push branches up got git. We CR-ed these branches before we merged them into our master, which cut down on bugs and helped make sure we all knew what was going on in our code base. 
What did your group do to try to keep your code DRY while many people collaborated on it? | We utilized view helpers and partials to keep things dry. We also assigned work with this in mind so we ran into fewer issues.
What was a technical challenge that you faced as a group? | The conditional validations on orders that showed/didn't show personal information on the site as appropriate.
What was a team/personal challenge that you faced as a group? | You only gave us three people. 
What could your team have done better? | Fine, we could have had higher test coverage, but we're happy with our 91%.
What was your application's ERD? (include a link) | https://www.lucidchart.com/documents/edit/2292d0d1-d497-4a72-ba8f-1089ed986071/0?shared=true&existing=1&docId=2292d0d1-d497-4a72-ba8f-1089ed986071
What is your Trello URL? | https://trello.com/b/TSmAKxla/supermarket
What is the Heroku URL of your deployed application? | https://superheromarket.herokuapp.com/

### CODE REVIEW

File: app/controllers/products_controller.rb [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+      flash[:failure] = 'You are not authorized to retire this product'
+    end
+    redirect_to merchant_manage_products_path(session[:merchant_id])
+  end
+
+  private
+
+  def product_params
+    return params.require(:product).permit(:name, :merchant_id, :stock, :price, :description, :photo_url, :category_ids => [])
+  end
```
Comment (droberts-sea) [inline app/controllers/products_controller.rb:81]: Product params should not include the merchant ID - you should be taking that from the session! I was able to change the owner of a product, effective "stealing" it.

File: app/controllers/products_controller.rb [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      render :new, status: :bad_request
+    end
+  end
+
+  def show; end
+
+  def edit; end
+
+  def update
+    @product.assign_attributes(product_params)
```
Comment (droberts-sea) [inline app/controllers/products_controller.rb:47]: Edit and update should both be making sure that the logged-in user is also the owner of the product. I was able to change the info on someone else's product by typing the URL into the address bar.

File: app/controllers/products_controller.rb [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+      redirect_to merchant_manage_products_path(session[:merchant_id])
+    else
+      flash.now[:failure] = 'Product not updated'
+      render :edit, status: :bad_request
+    end
+  end
+
+  def destroy
+    if @product.merchant_id == @logged_merchant.id
+
```
Comment (droberts-sea) [inline app/controllers/products_controller.rb:63]: Here you're checking for product ownership - good work. Since you need it in multiple places, this would be a good candidate for a controller filter.

File: app/models/merchant.rb [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    order_items = []
+    self.products.each { |product|
+      product.order_items.each { |order_item|
+        order_items << order_item
+      }
+    }
+
+    order_items.each { |order_item|
+      order_status = Order.find_by(id: order_item.order_id).status
+      if order_status == status || status == "all"
```
Comment (droberts-sea) [inline app/models/merchant.rb:25]: On line 24, why not `order_item.order.status`?

File: app/models/merchant.rb [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  validates :username, presence: true, uniqueness: true
+
+  validates :email, presence: true, uniqueness: true
+
+  def merchant_order_items(status)
+
+    orders_and_items = {}
+    orders = self.orders
+    order_ids = orders.map { |order| order.id }
+    order_ids.each { |order_id| orders_and_items[order_id] = [] }
```
Comment (droberts-sea) [inline app/models/merchant.rb:14]: I'm not sure what this method is accomplishing. Why not start with `orders.where(status: status)`, and then for each of those orders say `order.order_items`? Seems like putting it all into a hash of arrays is redundant.

File: app/models/merchant.rb [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+      end
+    end
+
+    return orders_and_items
+
+  end # merchant_order_items
+
+  def total_revenue_by(status)
+    orders_and_items = self.merchant_order_items("all")
+    total_revenue = 0.00
```
Comment (droberts-sea) [inline app/models/merchant.rb:42]: Since you're calling `merchant_order_items` here, why not pass in the status? That would save you the check on line 45.

File: app/models/merchant.rb [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    end
+    return total_revenue
+  end # total_revenue_by
+
+  def count_orders_by(status)
+    count = 0
+    orders = self.orders.distinct
+    orders.each do |order|
+      if order.status == status || status == "all"
+        count += 1
```
Comment (droberts-sea) [inline app/models/merchant.rb:58]: You shouldn't ever need to call `.distinct` on the collection you get from a relation. This whole method could be `self.orders.where(status: status).count`

File: app/models/order.rb [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    purchased_items.each do |item|
+      total += item.subtotal
+    end
+    return total
+  end
+
+  def order_status
+    statuses = self.order_items.map do |item|
+      item.status
+    end
```
Comment (droberts-sea) [inline app/models/order.rb:35]: Having both `order.status` and `order.order_status` is a little confusing. It might be wise to remove this column from the DB and have this be calculated on the fly every time. If that doesn't work, you could make it run automatically on an order by adding a `before_save` action on `OrderItem`.

File: app/models/order_item.rb [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    quantity = self.quantity
+    subtotal = (unit_price * quantity).round(2)
+    return subtotal
+  end
+
+  def self.existing_oi?(new_oi)
+    existing_oi = find_by(product_id: new_oi.product_id, order_id: new_oi.order_id)
+
+    return existing_oi.nil? ? nil : existing_oi
+  end
```
Comment (droberts-sea) [inline app/models/order_item.rb:30]: On line 29, you could just `return existing_oi`.

File: app/models/order.rb [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  validates :shipping_address, presence: true, on: :update
+  validates :billing_address, presence: true, on: :update
+
+  validates :order_items, length: { minimum: 1 }, if: :customer_info?
+
+  def customer_info?
+    if
+      self.customer_name.nil? || self.customer_email.nil? || self.credit_card.nil? || self.cvv.nil? || self.cc_expiration.nil? || self.shipping_address.nil? || self.billing_address.nil?
+      return false
+    end
```
Comment (droberts-sea) [inline app/models/order.rb:19]: This is exactly what we were talking about in the Unit 2 assessment review. The big long sequence of ors in the conditional will already evaluate to true or false, so you don't need the `if` at all, just `return self.customer_name.nil? || self.customer_email.nil? || ...`

File: app/controllers/merchants_controller.rb [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    merchant_id = params[:id]
+    @merchant = Merchant.find_by(id: merchant_id)
+    @products = Product.where(retired: false, merchant_id: @merchant.id)
+  end
+
+  # for the merchant
+  def edit
+  end
+
+  def update
```
Comment (droberts-sea) [inline app/controllers/merchants_controller.rb:19]: I wasn't able to get edit or update to work - you should fix them or delete these methods.

File: app/controllers/orderitems_controller.rb [hunk 12]
```diff
@@ -0,0 +1,86 @@
+class OrderitemsController < ApplicationController
+  before_action :find_order_item, only: [:update, :destroy, :ship, :cancel]
+  before_action :order_exists?, only: [:create]
+
+
+  def create
+    @orderitem = OrderItem.new(order_item_params)
+    @orderitem.order_id = session[:cart_id]
+    # order = Order.find(session[:cart_id])
```
Comment (droberts-sea) [inline app/controllers/orderitems_controller.rb:9]: I like that you've used model methods to hide some of the details here, but I think you could go even further. What about something like:

```ruby
order = Order.find(session[:cart_id])
@order_item = order.add_or_update(order_item_params)
if @order_item.save
```

`add_or_update` would do all the work that currently happens here.

File: app/controllers/orderitems_controller.rb [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    @orderitem.destroy
+    flash[:success] = 'Item successfully removed'
+
+    redirect_back(fallback_location: root_path)
+  end
+
+  def ship
+    @orderitem.assign_attributes(status: "shipped")
+    if @orderitem.save
+      flash[:success] = "You have shipped #{@orderitem.product.name} for order #{@orderitem.order.id}."
```
Comment (droberts-sea) [inline app/controllers/orderitems_controller.rb:48]: You should probably make sure that the person who's currently logged in is also the owner of this product.

File: config/routes.rb [hunk 14]
```diff
@@ -0,0 +1,37 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+  root 'products#home'
+
+  resources :orders
+
+  resources :merchants, except: [:new, :create] do
+    resources :orders, only: [:index]
+    resources :products, except: [:show]
```
Comment (droberts-sea) [inline config/routes.rb:10]: I'm not sure any of the `product` routes other than `index` need to be nested. If you need the merchant, you should either be looking it up in the session, or pulling it off the product itself.

File: test/controllers/categories_controller_test.rb [hunk 15]
```diff
@@ -0,0 +1,29 @@
+require "test_helper"
+
+describe CategoriesController do
+  describe'create' do
+    let(:merchant) { Merchant.first }
+    before do
+      login(merchant)
```
Comment (droberts-sea) [inline test/controllers/categories_controller_test.rb:7]: If creating a category requires you to be logged in, you should include a test where a guest user creates one and is rejected.

File: test/controllers/orderitems_controller_test.rb [hunk 16]
```diff
@@ -0,0 +1,178 @@
+require "test_helper"
+
+describe OrderitemsController do
+
+  describe 'create' do
+    it 'can add a valid orderitem' do
+      oi_data = {
```
Comment (droberts-sea) [inline test/controllers/orderitems_controller_test.rb:7]: I would include separate tests here for if the cart already exists or if it should create a new cart

File: test/controllers/orderitems_controller_test.rb [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+      must_redirect_to viewcart_path(session[:cart_id])
+      test_io.reload
+      test_io.quantity.must_equal 2
+
+    end
+
+    it 'sends does not update a product for invalid data' do
+      orderitem = {
+        product_id: Product.first.id,
+        order_id: Order.first.id,
```
Comment (droberts-sea) [inline test/controllers/orderitems_controller_test.rb:90]: What is this test name?

File: test/controllers/orderitems_controller_test.rb [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+
+      must_redirect_to root_path
+      OrderItem.count.must_equal old_oi_count
+    end
+  end
+
+  describe 'ship' do
+    it "allows you to ship an item that exists" do
+      order_item = OrderItem.create(order_id: Order.first.id, product_id: Product.first.id, quantity: 1, status: "pending")
+      order_item.status.must_equal "pending"
```
Comment (droberts-sea) [inline test/controllers/orderitems_controller_test.rb:143]: I would probably include some test cases around auth here - what if you're not logged in? What if you're logged in as someone other than the merchant offering this product?

File: test/controllers/products_controller_test.rb [hunk 19]
```diff
@@ -0,0 +1,228 @@
+require "test_helper"
+
+describe ProductsController do
+  describe 'guest user' do
+    describe 'index' do
+      it 'succeeds with multiple products for a guest user' do
```
Comment (droberts-sea) [inline test/controllers/products_controller_test.rb:6]: You need to include negative cases for the things a guest user / the wrong user can't do.

File: test/controllers/reviews_controller_test.rb [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+      post reviews_path, params: {review: review_data}
+
+      must_respond_with :redirect
+      Review.count.must_equal @old_review_count
+    end
+
+    it "wont' let you review if it's your own product" do
+      merchant = Merchant.first
+      merchant.products.length.must_be :>, 0
+      login(merchant)
```
Comment (droberts-sea) [inline test/controllers/reviews_controller_test.rb:36]: Good! More of this sort of test.

File: test/models/merchant_test.rb [hunk 21]
```diff
@@ -0,0 +1,219 @@
...
+        result = @merchant.total_revenue_by("completed")
+        result.must_equal completed_orders_total
+
+      end
+
+
+      it "returns 0.0 for the total revenue of a merchant with zero orders" do
+        merchant = Merchant.new
+        result = merchant.total_revenue_by("all")
+        result.must_equal 0.0
```
Comment (droberts-sea) [inline test/models/merchant_test.rb:159]: Good catch of this edge case

**Comment by droberts-sea** [comment]
# bEtsy
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** | 
Appropriate Git Usage with all members contributing | yes - I'm glad that code reviews worked well for you
Answered comprehension questions | yes
Trello board is created and utilized in project management | yes
Heroku instance is online | yes
**General** |
Nested routes follow RESTful conventions | mostly
oAuth used for User authentication | yes - good work getting Google sorted out
Functionality restricted based on user roles | some - see inline (products controller)
Products can be added and removed from cart | yes
Users can view past orders | yes
Merchants can add, edit and view their products | yes
Errors are reported to the user | yes
**Order Functionality** |
Reduces products' inventory | yes
Cannot order products that are out of stock | yes
Changes order state | yes
Clears current cart | yes
**Database** |
ERD includes all necessary tables and relationships | yes
Table relationships | yes
**Models** |
Validation rules for Models | yes
Business logic is in the models | mostly - some places it's not enough, some places it's overdone / overcomplicated
**Controllers** |
Controller Filters used to DRY up controller code | yes
**Testing** |
Model Testing | yes
Controller Testing | mostly - Many test cases missing around authorization. Since these mostly rely on controller filters they won't show up in simplecov, but they're still important! See inline for examples.
Session Testing | yes
SimpleCov at 90% for controller and model tests | yes - great job!
**Front-end** |
The app is styled to create an attractive user interface | yes - that contrast is a little hard to swallow, but it fits the theme and makes for a consistent look and feel
Content is organized with HTML5 semantic tags | yes
CSS is DRY | yes
**Overall** | Great work overall! You've built a fully functional web store from top to bottom, everything short of payment processing. This represents a huge amount of work, and you should be proud of yourselves! Your code is pretty well organized all the way from the CSS to the database, and I am quite happy with this submission. There are definitely places where things could be cleaned up or refined, or where there are feature gaps you might like to fill - I've tried to outline some of these below. However bEtsy is a huge project on a very short timeline, and this feedback should not at all diminish the magnitude of what you've accomplished. Keep up the hard work!

# Only the person who submitted the PR will get an email about this feedback. Please let the rest of your team know about it.
